### PR TITLE
Add example profiler that does stack sampling with ICorProfilerInfo10::SuspendRuntime

### PR DESCRIPTION
**Pull request by davmason**

### CODE REVIEW

**Comment by sdmaclea** [comment]
Why are we doing this in the dotnet/runtime repo?  It seems the correct place for sample code is in the dotnet/samples repo.

**Comment by davmason** [comment]
> 
> 
> Why are we doing this in the dotnet/runtime repo? It seems the correct place for sample code is in the dotnet/samples repo.

I view that as more of C# samples developers. I don't mind if code lives over there, but it doesn't seem like it belongs there any more than here.

This is a repeated problem the profiler runs in to. It's pretty different from anything else we do so there's no clear home for a lot of documentation/samples/etc.

**Comment by davmason** [comment]
Ping @noahfalk

File: docs/design/coreclr/profiling/sample-profilers/stacksampling/src/CorProfiler.cpp [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+HRESULT STDMETHODCALLTYPE CorProfiler::DynamicMethodJITCompilationFinished(FunctionID functionId, HRESULT hrStatus, BOOL fIsSafeToBlock)
+{
+    return S_OK;
+}
+
+bool CorProfiler::IsRuntimeExecutingManagedCode()
+{
+    return jitEventCount.load() > 0;
```
Comment (noahfalk) [inline docs/design/coreclr/profiling/sample-profilers/stacksampling/src/CorProfiler.cpp]: The runtime can execute managed code from a R2R image that doesn't involve jitting. Is there a reason we can't call SuspendRuntime() without first doing this check?

**Review by noahfalk — APPROVED** [review]
LGTM other than the jit event tracking

File: docs/design/coreclr/profiling/sample-profilers/stacksampling/src/CorProfiler.cpp (hunk 1 above)
Comment (davmason) [reply docs/design/coreclr/profiling/sample-profilers/stacksampling/src/CorProfiler.cpp]: You can suspend the runtime anytime after g_fEEStarted is set. This bit of code isn't a functional requirement, it just takes debug runtimes forever to start executing managed code and I wanted to wait to sample until there was something to sample so the logs would be cleaner.

**Comment by jkotas** [comment]
> I view that as more of C# samples developers.

C# samples have advanced C++ samples too, e.g.: https://github.com/dotnet/samples/tree/master/core/hosting

I agree with @sdmaclea that this would be better to have in dotnet/samples.

> It's pretty different from anything else we do so there's no clear home for a lot of documentation/samples/etc

The profiler APIs docs have a home in the docs, e.g. :
https://github.com/dotnet/docs/blob/master/docs/framework/unmanaged-api/profiling/profiling-interfaces.md (https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo-interface is rendeded view of this content.)

Have you talked about this to the docs team?

**Comment by davmason** [comment]
> I agree with @sdmaclea that this would be better to have in dotnet/samples.

 I don't have super strong feelings about it, so I can move it. I mainly just want it to be somewhere so I can point people to it when they ask. I didn't see native hosting samples.

> Have you talked about this to the docs team?

We have had some preliminary discussions about updating the API docs you referred above to reflect the growing differences in core and desktop. The documentation I was talking about was the long form, blog post style docs in the docs/design/coreclr/profiling path. I haven't had discussions with them about this style of documentation.

**Comment by jkotas** [comment]
>  The documentation I was talking about was the long form, blog post style docs 

This type of documentation is there as well, e.g. https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/profiling-overview

**Pull request merged**